uaclient/apt_news.py

```diff
@@ -0,0 +1,180 @@
+import datetime
+import json
+import logging
+import unicodedata
+from typing import List, Optional
+
+import apt_pkg
+
+from uaclient import defaults, system, util
+from uaclient.clouds.identity import get_cloud_type
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    DatetimeDataValue,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.files import state_files
+
+
+class AptNewsMessageSelectors(DataObject):
+    fields = [
+        Field("codenames", data_list(StringDataValue), required=False),
+        Field("clouds", data_list(StringDataValue), required=False),
+        Field("pro", BoolDataValue, required=False),
+    ]
+
+    def __init__(
+        self,
+        *,
+        codenames: Optional[List[str]] = None,
+        clouds: Optional[List[str]] = None,
+        pro: Optional[bool] = None
+    ):
+        self.codenames = codenames
+        self.clouds = clouds
+        self.pro = pro
+
+
+class AptNewsMessage(DataObject):
+    fields = [
+        Field("begin", DatetimeDataValue),
+        Field("end", DatetimeDataValue, required=False),
+        Field("selectors", AptNewsMessageSelectors, required=False),
+        Field("lines", data_list(StringDataValue)),
+    ]
+
+    def __init__(
+        self,
+        *,
+        begin: datetime.datetime,
+        end: Optional[datetime.datetime] = None,
+        selectors: Optional[AptNewsMessageSelectors] = None,
+        lines: List[str]
+    ):
+        self.begin = begin
+        self.end = end
+        self.selectors = selectors
+        self.lines = lines
+
+
+def format_message(msg: AptNewsMessage) -> str:
+    result = "#\n"
+    for line in msg.lines:
+        result += "# {}\n".format(line)
+    result += "#\n"
+    return result
+
+
+def do_selectors_apply(
+    cfg: UAConfig, selectors: Optional[AptNewsMessageSelectors]
+) -> bool:
+    if selectors is None:
+        return True
+
+    if selectors.codenames is not None:
+        if system.get_platform_info()["series"] not in selectors.codenames:
+            return False
+
+    if selectors.clouds is not None:
+        cloud_id, fail = get_cloud_type()
+        if fail is not None:
```

Comment:
Should we log here that `cloud-id` failed somehow ?

Reply:
We could do that, though I suggest we put it inside the get_cloud_type function like this:

```diff
 @lru_cache(maxsize=None)
 @apply_config_settings_override("cloud_type")
 def get_cloud_type() -> Tuple[Optional[str], Optional[NoCloudTypeReason]]:
     if system.which("cloud-id"):
         # Present in cloud-init on >= Xenial
         try:
             out, _err = system.subp(["cloud-id"])
             return (out.strip(), None)
-        except exceptions.ProcessExecutionError:
+        except exceptions.ProcessExecutionError as e:
+            logging.debug("error running cloud-id: %s", str(e))
             return (None, NoCloudTypeReason.CLOUD_ID_ERROR)
     # If no cloud-id command, assume not on cloud
     return (None, NoCloudTypeReason.NO_CLOUD_DETECTED)
```

I can add that in an extra commit if you agree

Reply:
Fair enough, that's a better change for sure

Reply:
added!